crates/llm-base/src/vocabulary.rs

```diff
@@ -18,9 +19,115 @@ pub enum TokenizationError {
+    fn tokenize<'a>(
+        &'a self,
+        text: &str,
+        bos: bool,
+    ) -> Result<Vec<(Vec<u8>, TokenId)>, TokenizationError>;
+}
+
+/// Vocabulary enum
+pub enum Vocabulary {
+    /// The vocabulary built-in to the model.
+    Ggml(GgmlVocabulary),
+
+    /// A custom vocabulary provided by the user.
+    Tokenizer(TokenizerVocabulary),
+}
+
+impl Vocabulary {
+    /// Create a new vocabulary with the default GGML vocabulary.
+    pub fn new_ggml() -> Self {
+        Vocabulary::Ggml(GgmlVocabulary::default())
+    }
+
+    /// Create a new vocabulary with a custom tokenizer.
+    pub fn new_tokenizer(tokenizer: Tokenizer) -> Self {
+        Vocabulary::Tokenizer(TokenizerVocabulary::new(tokenizer))
+    }
+
+    /// Add a token to the vocabulary.
+    ///
+    /// The token added must have `id` directly after the last token in the vocabulary.
+    ///
+    /// # Panics
+    /// - This function can panic if `id` does not correspond to the next token in the vocabulary.
+    ///   That is, if there are already `n` tokens in the vocabulary, then `id` must be `n`.
+    pub fn push_token(&mut self, id: TokenId, content: Token, score: TokenScore) {
+        match self {
+            Vocabulary::Ggml(v) => v.push_token(id, content, score),
+            Vocabulary::Tokenizer(v) => v.push_token(id, content, score),
+        }
+    }
+
+    /// Converts a token to the token ID it represents in this vocabulary.
+    pub fn token_to_id(&self, token: &[u8]) -> Option<TokenId> {
+        match self {
+            Vocabulary::Ggml(v) => v.token_to_id(token),
+            Vocabulary::Tokenizer(v) => v.token_to_id(token),
+        }
+    }
+
+    /// Converts a token index to the token it represents in this vocabulary.
+    pub fn token(&self, idx: usize) -> Vec<u8> {
+        match self {
+            Vocabulary::Ggml(v) => v.token(idx),
+            Vocabulary::Tokenizer(v) => v.token(idx),
+        }
+    }
+
+    /// Returns the number of tokens in the vocabulary.
+    pub fn len(&self) -> usize {
+        match self {
+            Vocabulary::Ggml(v) => v.len(),
+            Vocabulary::Tokenizer(v) => v.len(),
+        }
+    }
+
+    /// Returns whether the vocabulary is empty.
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Vocabulary::Ggml(v) => v.is_empty(),
+            Vocabulary::Tokenizer(v) => v.is_empty(),
+        }
+    }
+
+    /// Tokenize a `text` with this vocabulary.
+    ///
+    /// `bos` controls whether a beginning-of-string token should be inserted.
+    pub fn tokenize<'a>(
+        &'a self,
+        text: &str,
+        bos: bool,
+    ) -> Result<Vec<(Vec<u8>, TokenId)>, TokenizationError> {
+        match self {
+            Vocabulary::Ggml(v) => v.tokenize(text, bos),
+            Vocabulary::Tokenizer(v) => v.tokenize(text, bos),
+        }
+    }
+}
+
+/// The built-in GGML vocabulary.
 #[derive(Debug, Clone, Default)]
-pub struct Vocabulary {
+pub struct GgmlVocabulary {
```

Comment:
Should this be called `ModelVocabulary` since we're actively trying to not be tightly coupled with GGML?